old/soft3/smoothtriang/soft/SF/man/man3/mesh.3
.TH mesh 3
.SH NAME
MeshParse, WriteMesh, WriteItoEMesh, BoundaryEdge, BoundaryVertex,
BoundaryFace, ForeachMeshFace, ForeachMeshVertex, ForeachMeshEdge,
ForeachFaceVertex, ForeachFaceEdge,
ForeachVertexFace, ForeachVertexEdge, ForeachVertexVertex, 
EndForeach, GetEdgeVertices, GetVertexEdge, GetEdgeFaces,
SetInternalData,ReturnInternalData,SetName,ReturnName
\- mesh operations
.SH SYNTAX
.nf
.B #include "geometry.h"
.B #include "userData.h"
.B #include <mesh.h>
.PP
.B Mesh* MeshParse(fp)
.B FILE* fp;
.PP
.B void WriteMesh(m,fp)
.B Mesh* m;
.B FILE* fp;
.PP
.B void WriteItoEMesh(m, fp, itoeV, itoeE, itoeF)
.B Mesh* m;
.B FILE* fp;
.B void* (*itoeV)();
.B void* (*itoeE)();
.B void* (*itoeF)();
.PP
.B void ClearMeshExternalData(m)
.B Mesh* m;
.PP
.B BOOLEAN BoundaryEdge(e)
.B Edge* e;
.PP
.B BOOLEAN BoundaryVertex(v)
.B Vertex* v;
.PP
.B BOOLEAN BoundaryFace(f)
.B Face* f;
.PP 
.B Vertex* VertexPath(v1,v2,p)
.B Vertex* v1;
.B Vertex* v2;
.B char* p;
.PP
.B ForeachMeshFace(m,f)
.B Mesh* m;
.B Face* f;
.PP 
.B ForeachMeshVertex(m,v)
.B Mesh* m;
.B Vertex* v;
.PP 
.B ForeachMeshEdge(m,e)
.B Mesh* m;
.B Edge* e;
.PP 
.B ForeachFaceVertex(f,v)
.B Face* f;
.B Vertex* v;
.PP 
.B ForeachFaceEdge(f,e)
.B Face* f;
.B Edge* e;
.PP 
.B ForeachFaceFace(f1,f2)
.B Face* f1;
.B Face* f2;
.PP 
.B ForeachVertexFace(v,f)
.B Vertex* m;
.B Face* f;
.PP 
.B ForeachVertexEdge(v,e)
.B Vertex* m;
.B Edge* f;
.PP 
.B ForeachVertexVertex(v1,v2)
.B Vertex* v1;
.B Vertex* v2;
.PP 
.B void GetVertexEdge(v1,v2,e)
.B Vertex* v1;
.B Vertex* v2;
.B Edge** e;
.PP
.B GetEdgeVertices(e,v1,v2)
.B Edge* e;
.B Vertex** v1;
.B Vertex** v2;
.PP
.B GetEdgeFaces(e,f1,f2)
.B Edge* e;
.B Face** f1;
.B Face** f2;
.PP
.B GetFaceEdge(f1,f2,e)
.B Face* f1;
.B Face* f2;
.B Edge** e;
.PP
.B GetVertexFaceEdge(v, f, e)
.B Vertex* v;
.B Face* f;
.B Edge** e;
.PP
.B SetInternalData(v, d)
.B Vertex* v;
.B void* d;
.PP
.B void* ReturnInternalData(v)
.B Vertex* v;
.PP
.B SetName(v, n)
.B Vertex* v;
.B char* n;
.PP
.B ReturnName(v, n)
.B Vertex* v;
.B char* n;
.fi
.SH DESCRIPTION
Mesh is a library to manipulate topological data structures.
Currently, meshes can be read in, traversed, and written.
The mesh package does NOT handle any geometry.  A separate
package is provided to allow certain geometric operations.
See userdata(3).

The mesh library also doesn't allow for the modification of
meshes.  This is handled in the library operators(3).  Note
that some of the iterators below will have problems if the
mesh is modified in the middle of an iteration.

.PP
The
.B MeshParse()
routine reads in a mesh from a file, and returns a pointer to
a mesh structure.  Note that this "routine" is really a macro.

Each time MeshParse() (or another from its family) is called, the
next mesh in the file is read.  Note that each point and faces may
only occur in one mesh.  Any point or face not used in a mesh will
remain defined for the next mesh that is read.

See also MeshUDParse (userdata 3).
.PP
The
.B WriteMesh()
routine writes a mesh to the given file.  Note that if the internal
data fields have been modified, then a call to ConvertGeometryToExternal()
should be made to flush them to the dstructs.
A more memory efficient approach is to call
.B WriteItoEMesh().
The routine takes three procedures that convert the internal format
to the external format while writing, free the external space after
is is printed.  See userdata 3 for a default set of 
.B itoe 
routines.

.B ClearMeshExternalData()
clears the external data fields of a mesh.
.PP
The routines
.B BoundaryEdge(), BoundaryVertex(),
and
.B BoundaryFace()
return TRUE if the parameter is an object of the appropriate type
that lies on the boundary of the mesh.  A face is said to lie
on the boundary if any of its edges or vertices lie on the
boundary.
.PP
The routine
.B VertexPath()
walks a path through the mesh.  The path starts at
.I v1.
It next moves to 
.I v2.
This gives us a vertex and an edge (the edge from
.I v1
to
.I v2
).  Our path will be a string of integers, separated by spaces.
From 
.I v2,
we will label the edges in a counter clockwise fashion, giving the
label 0 to the edge from
.I v1
to 
.I v2.
At each successive vertex in the path, the edge the vertex was entered
from will be labeled 0.  By specifing negative numbers, one may move
in the clockwise direction.  If a number is larger than the number of
vertices in the path, then the count "wraps".  Boundaries are skipped
over.  If
.I v2
is not adjacent to
.I v1,
a NULL pointer is returned.  Otherwise, the last vertex in the path
is returned.
.PP
The "routines"
.B ForeachMeshFace(), ForeachMeshVertex(), ForeachMeshEdge(),
.B ForeachVertexFace(), ForeachVertexFace(), ForeachFaceVertex(), 
.B ForeachFaceEdge(), ForeachFaceFace(),
and
.B ForeachVertexVertex()
are macros that allow iteration over various parts of the mesh.
The type of iteration is determined by the name: the first part
is what is iterated around, while the second part is what is
iterated over.  For example, 
.B ForeachMeshVertex()
will give each vertex in the mesh.  
Each loop is terminated by the macro
.B EndForeach.
Iterators which iterator around an object in the mesh will iterate
in counter-clockwise order.  The current object will be stored in
the second parameter to the 
.B Foreach.
The first object in the iteration is arbitrary, unless the object
iterated around is on the boundary.  In this case, the iteration
will start at one part of the boundary and end at the other part.
A local variable,
.B iter,
is used as a loop control variable.

The routine
.B GetEdgeVertices()
gets both of the vertices associated with an edge.
The routine
.B GetVertexEdge()
finds the directed edge associated with two vertices.
If the edge doesn't exist, the edge pointer is set to NULL.
.B GetEdgeFaces()
gets the two faces neighboring an edge.  If the edge is a boundary
edge, then the second face will be set to NULL.
.B GetFaceEdge()
finds the edge between the two faces.  If the edge doesn't exist,
the edge pointer is set to NULL.
.B GetVertexFaceEdge()
finds the edge bounding f leaving v in a counter clockwise direction.
If v is not on f, then the edge pointer will be set to NULL.

NOTE: If you forget an EndForeach, you will get strange error
messages.  Most commonly, you'll get a syntax error at the
start of the next procedure or at the end of the file.

NOTE: ForeachVertexEdge returns all the edges pointing to a particular
vertex.  This means that, for boundary vertices, the boundary half edge 
directed away from the vertex will not be included in the iteration.

.SH EXAMPLE
The following code reads in a mesh, and prints out the names of
all the faces followed by the ring of vertices around each face.

.EX
	m = MeshParse(stdin);
	ForeachMeshFace(m,f)
		printf("face %s: ",ReturnName(f));
		ForeachFaceVertex(f,v)
			printf(" %s ",ReturnName(v));
		EndForeach
		printf("\\n");
	EndForeach
.EE

.SH NOTES
This is an early version, and is subject to change.
.PP
Send bugs/requests to smann@june.

.SH SEE ALSO
userdata(3), operators(3)

.SH DATE
November 10, 1991


old/soft3/smoothtriang/soft/SF/src/Mesh/Mesh/mesh.c
/*
 * Copyright (c) 1990, Graphics and AI Laboratory, University of Washington
 * Copying, use and development for non-commercial purposes permitted.
 *                  All rights for commercial use reserved.
 */

/*
 *----------------------------------------------------------------------
 *  File:  mesh.c
 *  Last Modified: Tue Jun 27, 1989 at 07:43:54 AM
 *----------------------------------------------------------------------
 */

#include <stdio.h>
#include <strings.h>
#include <ctype.h>
#include "grailTypes.h"
#include "geometry.h"
#include "dstruct.h"
#include "mesh.h"
#include "userData.h"
#include "meshsym.h"


/* ========= FLAGS =========== */

/*
 *----------------------------------------------------------------------
 *  Function:  SetIFlags
 *----------------------------------------------------------------------
 */
void SetIFlags(Node* n, int f)
{
	n->iflags |= f;
}

void ClearIFlags(Node* n, int f)
{
	n->iflags &= ~f;
}

int QIFlags(Node* n, int f)
{
	return (n->iflags & f);
}


/* ========= PREDICATES ========= */

/*
 *----------------------------------------------------------------------
 *  Function:  BoundaryEdge
 *	Return TRUE if an edge is a boundary edge.
 *----------------------------------------------------------------------
 */
BOOLEAN BoundaryEdge(Edge* e)
{
  if ( e == NULL  ||  e->sym == NULL )
    return TRUE;
  else return FALSE;
}



/*
 *----------------------------------------------------------------------
 *  Function:  BoundaryVertex
 *	Return TRUE if a vertex is a boundary vertex.
 *----------------------------------------------------------------------
 */
BOOLEAN BoundaryVertex(Vertex* v)
{
  Edge* e;

  if ( v->tag != POINT ){
    fprintf(stderr,"BoundaryVertex(): %s not a point.  Exiting.\n",v->name);
    exit(1);
  }
  if ( v->rep == NULL )
    return TRUE;
  e = v->rep;
  do {
    if ( e->next == NULL  ||  e->next->sym == NULL )
      return TRUE;
    e = e->next->sym;
  } while ( e != v->rep );
  return FALSE;
}




/*
 *----------------------------------------------------------------------
 *  Function:  BoundaryFace
 *	Return TRUE if a face is a boundary face.
 *  A face is a boundary face if any of it's edges or vertices
 *  lie on the boundary.
 *----------------------------------------------------------------------
 */
BOOLEAN BoundaryFace(Face* f)
{
  Edge* e;
  Vertex* v;

  if ( f->tag != FACE ){
    fprintf(stderr,"BoundaryFace(): %s is not a face.  Exiting.\n",f->name);
    exit(1);
  }

  e = f->rep;
  do {
    if ( e->sym == NULL  ||  e->next == NULL )
      return TRUE;
    e = e->next;
  } while ( e != f->rep );

  ForeachFaceVertex(f,v){
    if ( BoundaryVertex(v) )
      return TRUE;
  } EndForeach;

  return FALSE;
}






/*
 *----------------------------------------------------------------------
 *  Function:  GetEdgeVertices
 *----------------------------------------------------------------------
 */
void GetEdgeVertices(Edge* e, Vertex** v1, Vertex** v2)
{
  *v1 = e->prev->p;
  *v2 = e->p;
}


/*
 *----------------------------------------------------------------------
 *  Function:  GetVertexEdge
 *	Get the directed edge from one vertex to another.
 *  Parameters:
 *	Vertex *v1,*v2		- first and second vertices
 *	Edge** e		- the edge from v1 to v2
 *  Return Value:
 *	void
 *----------------------------------------------------------------------
 */
void GetVertexEdge(Vertex *v1,Vertex *v2, Edge **e )
{
	Edge* ei;
	Edge* ve=NULL;
	Vertex* u;
	Vertex* v;
	
	ForeachVertexEdge(v2,ei){
		GetEdgeVertices(ei,&u,&v);
		if ( u == v1 ){
			ve = ei;
		}
	} EndForeach;
	
	*e = ve;
}

/*
 *----------------------------------------------------------------------
 *  Function:  GetFaceEdge
 *	Get the directed edge between two faces, bounding face f1.
 *----------------------------------------------------------------------
 */
void GetFaceEdge(Face* f1, Face* f2, Edge** e)
{
	Edge* ei;
	Edge* fe=NULL;
	Face* u;
	Face* v;
	void GetEdgeFaces();

	ForeachFaceEdge(f1, ei) {
		if ( ei->sym != NULL ) {
			GetEdgeFaces(ei, &u, &v);
			if ( v == f2 ) {
				fe = ei;
			}
		}
	} EndForeach;

	*e = fe;
}


/*
 *----------------------------------------------------------------------
 *  Function:  GetVertexFaceEdge
 *	Get the edge shared by both v and f that runs in 
 *  counter-clockwise direction from v around f.
 *----------------------------------------------------------------------
 */
void GetVertexFaceEdge(Vertex* v, Face* f, Edge** e)
{
	Edge* e1;

	*e = NULL;
	ForeachFaceEdge(f, e1) {
		if ( e1->p == v ) {
			*e = e1->next;
			return;
		}
	} EndForeach;
}


/*
 *----------------------------------------------------------------------
 *  Function:  GetEdgeFaces
 *	Get the two faces neighboring an edge.  If the edge is a
 *  boundary edge, the second of the two faces will be NULL.
 *----------------------------------------------------------------------
 */
void GetEdgeFaces(Edge* e, Face** f1, Face** f2)
{
	*f1 = e->d;
	if ( e->sym == NULL ) {
		*f2 = NULL;
	} else {
		*f2 = e->sym->d;
	}
}






printEdge(Edge* e)
{
  if ( e == NULL ){
    printf("null_edge");
    return;
  }
  printf("'%s' - ",e->p->name);
  if ( e->sym != NULL )
    printf("'%s'",e->sym->p->name);
  else
    printf("X");
}




static void CompleteVertexEdgeRing(Vertex* v, Edge* e1, Edge* e2)
{
  Edge* e;

  e = v->rep;
  while ( e->next->sym != NULL ) {
    e = e->next->sym;
  }
  e->next->sym = e1;
  e1->sym = e->next;
  e1->p = v;
  e1->next = e2;
  e1->prev = NULL;
  e1->d = NULL;

  while( e->sym != NULL ) {
    e = e->sym->prev;
  }
  e->sym = e2;
  e2->sym = e;
  e2->p = e->prev->p;
  e2->next = NULL;
  e2->prev = e1;
  e1->d = NULL;
}



/*
 *----------------------------------------------------------------------
 *  Function:  sscani
 *	Read an integer from a string consisting only of spaces
 *  and integers, returning the next space after the integer.
 *----------------------------------------------------------------------
 */
static char* sscani(char* s, int *i)
{
  int n;
  int sign;
  while( *s == ' ' ){
    s++;
  }
  if ( !isdigit(*s) && *s != '-' )
    return NULL;
#if 0
  n = sscanf(s,"%d",i);
  if ( n == 0 )
    return NULL;
  else {
    s++; /* we know there is at least one char; this eats the '-' */
    while( isdigit(*s) )
      s++;
    return s;
  }
#endif
  if ( *s == '-' ){
    sign = -1;
    s++;
    if ( !isdigit(*s) )
      return NULL;
  }
  else {
    sign = 1;
  }
  *i = *s++ - '0';
  while(isdigit(*s)){
    *i = 10* *i + *s++ - '0';
  }
  *i = *i * sign;
  return s;
}


/*
 *----------------------------------------------------------------------
 *  Function:  vPath
 *	Traverse the vertex path p starting at v from e.
 *  e is the half edge that points to v, and is numbered 0.
 *
 *  HACK ALERT:
 *	There is a subtle hack that makes things work.
 *  There is a problem when the vertex is a boundary vertex.
 *  For simplicity, we "complete" the vertex ring, making traversal
 *  easier.  However, the edge we pass in the recursive call could
 *  be one of the half edges we introduced to complete the ring.
 *  *BECAUSE WE ONLY LOOK AT THE SYM POINTER OF THIS EDGE AND NEVER
 *   USE IT IN ANY OTHER MANNER (IN THE RECURSIVE CALL* this code
 *  works.  Be careful of your use of the parameter 'e' when modifing 
 *  this code since it doesn't really exist.
 *----------------------------------------------------------------------
 */
static Vertex* vPath(Vertex* v, Edge* e, char* p)
{
  int n;
  Edge e1,e2;
  int i;

  p = sscani(p,&n);
  if ( p == NULL ) 
    return v;

  if ( BoundaryVertex(v) ) {
    CompleteVertexEdgeRing(v,&e1,&e2);
  } else { /* do this to make later code work */
    e1.sym = &e1;
    e2.sym = &e2;
  }
  if ( n >= 0 ) {
    for( e=e->sym, i=0; i<n; i++ ){
      e = e->prev->sym;
    }
    e1.sym->sym = NULL;
    e2.sym->sym = NULL;
    return vPath(e->p,e,p);
  } else /* n < 0 */ {
    for( e=e->sym, i=0; i>n; i-- ){
      e = e->sym->next;
    }
    e1.sym->sym = NULL;
    e2.sym->sym = NULL;
    return vPath(e->p,e,p);
  }
}



Vertex* VertexPath(Vertex* v1, Vertex* v2, char* path)
{
  Edge* e;
  Edge e1,e2;
  int i;

  /* check and see if path is well formed */
  for(i=0;i<strlen(path);i++){
    switch(path[i]){
    case ' ': case '0': case '1': case '2': case '3': case '4': 
    case '5': case '6': case '7': case '8': case '9': case '-':
      continue;
    default:
      fprintf(stderr,"VertexPath: path '%s' malformed.  Must contain ",path);
      fprintf(stderr,"'-', ' ', and digits only.  Exiting.\n");
      exit(1);
    }
  }

  /* find edge between v1 and v2 */
  if ( BoundaryVertex(v1) ) {
    CompleteVertexEdgeRing(v1,&e1,&e2);
  } else { /* do this to make later code work */
    e1.sym = &e1;
    e2.sym = &e2;
  }
  e = v1->rep;
  while( e->sym->p != v2 ) {
    e = e->sym->prev;
    /* if we looped around and didn't find v2, then v2 doesn't exist */
    if ( e == v1->rep ){
      return NULL;
    }
  }
  e = e->sym;

  e1.sym->sym = NULL;
  e2.sym->sym = NULL;

  return vPath( v2, e, path );
}


/*
 *----------------------------------------------------------------------
 *  Function:  SetInternalData
 *----------------------------------------------------------------------
 */
void SetInternalData(Vertex* v, void* d)
{
	v->internalData = (Data*)d;
}

/*
 *----------------------------------------------------------------------
 *  Function:  ReturnInternalData
 *----------------------------------------------------------------------
 */
void* ReturnInternalData(Vertex* v)
{
	return v->internalData;
}



/* ============= NAMES ============ */

static int curName=0;

static int numSuffix(char* s)
{
	int len;
	int i;

	len = strlen(s);
	for (i=len-1; i>=0; i--) {
		if ( !isdigit(s[i]) ) {
			break;
		}
	}
	if ( i+1 == len ) {
		return 0;
	} else {
		return atoi(&(s[i+1]));
	}
}


static int maxNum(Mesh* m)
{
	Face* f;
	Edge* e;
	Vertex* v;
	int n;
	int maxn=0;

	ForeachMeshFace (m, f) {
		if ( ReturnName(f) ) {
			n = numSuffix(ReturnName(f));
			if ( n > maxn ) {
				maxn = n;
			}
		}
	} EndForeach;
	ForeachMeshEdge (m, e) {
		if ( ReturnName(e) ) {
			n = numSuffix(ReturnName(e));
			if ( n > maxn ) {
				maxn = n;
			}
		}
	} EndForeach;
	ForeachMeshVertex (m, v) {
		if ( ReturnName(v) ) {
			n = numSuffix(ReturnName(v));
			if ( n > maxn ) {
				maxn = n;
			}
		}
	} EndForeach;

	return maxn;
}

/*
 *----------------------------------------------------------------------
 *  Function:  SetName
 *----------------------------------------------------------------------
 */
void SetName(void* p, char* n)
{
	Vertex* v=(Vertex*)p;
	if ( v->name != NULL ) {
		free(v->name);
	}
	if ( n != NULL ) {
		if ( numSuffix(n) > curName ) {
			curName = numSuffix(n) + 1;
		}
		v->name = (char*)malloc(strlen(n)+1);
		strcpy(v->name, n);
	} else {
		v->name = NULL;
	}
}

/*
 *----------------------------------------------------------------------
 *  Function:  ReturnName
 *----------------------------------------------------------------------
 */
char* ReturnName(void* v)
{
	return ((Vertex *)v)->name;
}


char* MeshUniqueName(Mesh* m, char* s)
{
	static char buf[100];
	static int first=1;

	/* we'll assume one mesh... */
	if ( first  ) {
		curName = maxNum(m)+1;
	}
	first = 0;

	if ( s == NULL  ||  *s == NULL ) {
		sprintf(buf, "X%d", curName++);
	} else {
		sprintf(buf, "%sX%d", s, curName++);
	}
	return buf;
}
